src/control/ros/RosJointStateClient.cpp

```diff
@@ -0,0 +1,106 @@
+#include <aikido/control/ros/RosJointStateClient.hpp>
+
+namespace aikido {
+namespace control {
+namespace ros {
+
+//=============================================================================
+RosJointStateClient::RosJointStateClient(
+      dart::dynamics::SkeletonPtr _skeleton,
+      ::ros::NodeHandle _nodeHandle,
+      const std::string& _topicName,
+      size_t capacity)
+  : mSkeleton{std::move(_skeleton)}
+  , mBuffer{}
+  , mCapacity{capacity}
+  , mCallbackQueue{} // Must be before mNodeHandle for order of destruction.
+  , mNodeHandle{std::move(_nodeHandle)}
+{
+  if (!mSkeleton)
+    throw std::invalid_argument("Skeleton is null.");
+
+  if (capacity < 1)
+    throw std::invalid_argument("Capacity must be positive.");
+
+  mNodeHandle.setCallbackQueue(&mCallbackQueue);
+  mSubscriber = mNodeHandle.subscribe(_topicName, 1,
+    &RosJointStateClient::jointStateCallback, this);
+}
+
+//=============================================================================
+void RosJointStateClient::spin()
+{
+  std::lock_guard<std::mutex> skeletonLock{mSkeleton->getMutex()};
+  std::lock_guard<std::mutex> bufferLock{mMutex};
+
+  mCallbackQueue.callAvailable();
+}
+
+//=============================================================================
+Eigen::VectorXd RosJointStateClient::getLatestPosition(
+  const dart::dynamics::MetaSkeleton& _metaSkeleton) const
+{
+  std::lock_guard<std::mutex> bufferLock{mMutex};
+  Eigen::VectorXd position(_metaSkeleton.getNumDofs());
+
+  for (size_t idof = 0; idof < _metaSkeleton.getNumDofs(); ++idof)
+  {
+    const auto dof = _metaSkeleton.getDof(idof);
+    const auto it = mBuffer.find(dof->getName());
+    if (it == std::end(mBuffer))
+    {
+      std::stringstream msg;
+      msg << "No data is available for '" << dof->getName() << "'.";
+      throw std::runtime_error(msg.str());
+    }
+
+    const auto& buffer = it->second;
+    if (buffer.empty())
+    {
+      std::stringstream msg;
+      msg << "Data for '" << dof->getName() << "' is invalid.";
+      throw std::runtime_error(msg.str());
+    }
+
+    const auto& record = buffer.back();
+    position[idof] = record.mPosition;
+  }
+
+  return position;
+}
+
+//=============================================================================
+void RosJointStateClient::jointStateCallback(
+  const sensor_msgs::JointState& _jointState)
+{
+  // This method assumes that mSkeleton->getMutex() and mMutex are locked.
```

Comment:
We don't lock here because `spin` acquires the mutex once before calling `callAvailable`. This eliminates the considerable overhead of locking the mutex multiple times if there are a large number of messages queued.

It also means that `spin()` is atomic - the return value of `getLatestPosition()` will _only_ change when `spin()` returns. This is not a huge benefit, but (IMO) the fewer concurrent changes that are happening the easier things are to debug.

> Isn't it safe to lock a locked mutex again?

Only recursive mutexes, which we are **not** using here. In my opinion (which is concisely summarized [here](https://cppwisdom.quora.com/Recursive-mutexes-considered-harmful)), recursive mutexes are generally a bad idea because they mask subtle race conditions.

As [the engineer who implemented the POSIX recursive mutex puts it](https://groups.google.com/forum/?hl=en#!topic/comp.programming.threads/tcrTKnfP8HI%5B51-75%5D):

>Really, really. Recursive mutexes are really bad and they really do hide serious bugs.
> [...]
> But if that's all that's necessary, why does POSIX have recursive mutexes? Because of a dare.

😈